doc/guides/using-internal-errors.md

```diff
@@ -0,0 +1,112 @@
+used to produce `Error`, `TypeError` and `RangeError` instances that use a
+static, permanent error code and an optionally parameterized message. The
+intent of the module is to make errors thrown by Node.js core to be more
+consistent and to allow changes to error messages to be made without requiring
+a semver-major label.
+
+Switching an existing error to use the `internal/errors` module must be
+considered a semver-major change. However, once using `internal/errors`,
+changes to `internal/errors` error messages will be handled as semver-minor
+or patch.
+
+## Using internal/errors.js
+
+The `internal/errors` module exposes three custom `Error` classes as a
+`message` function.
+
+### Class: errors.Error(key[, args...])
+
+* `key` {String} The static error identifier
+* `args... {Any} Zero or more optional arguments
+
+```js
+const errors = require('internal/errors');
+
+var arg1 = 'foo';
+var arg2 = 'bar';
+const myError = new errors.Error('KEY', arg1, arg2);
+throw myError;
+```
+
+The specific error message for the `myError` instance will depend on the
+associated value of `KEY` (see "Adding new errors").
+
+The `myError` object will have a `code` property equal to the `key` and a
+`name` property equal to `Error[${key}]`.
+
+### Class: errors.TypeError(key[, args...])
+
+* `key` {String} The static error identifier
+* `args... {Any} Zero or more optional arguments
+
+```js
+const errors = require('internal/errors');
+
+var arg1 = 'foo';
+var arg2 = 'bar';
+const myError = new errors.TypeError('KEY', arg1, arg2);
+throw myError;
+```
+
+The specific error message for the `myError` instance will depend on the
+associated value of `KEY` (see "Adding new errors").
+
+The `myError` object will have a `code` property equal to the `key` and a
+`name` property equal to `TypeError[${key}]`.
+
+### Class: errors.RangeError(key[, args...])
+
+* `key` {String} The static error identifier
+* `args... {Any} Zero or more optional arguments
+
+```js
+const errors = require('internal/errors');
+
+var arg1 = 'foo';
+var arg2 = 'bar';
+const myError = new errors.RangeError('KEY', arg1, arg2);
+throw myError;
+```
+
+The specific error message for the `myError` instance will depend on the
+associated value of `KEY` (see "Adding new errors").
+
+The `myError` object will have a `code` property equal to the `key` and a
+`name` property equal to `RangeError[${key}]`.
+
+### Method: errors.message(key[, args...])
+
+* `key` {String} The static error identifier
+* `args...` {Any} Zero or more optional arguments
+* Returns: {String}
+
+Returns the formatted error message string for the given `key`.
+
+## Adding new errors
+
+New static error codes are adding by modifying the `internal/errors.js` file
+and appending the new error codes to the end using the utility `E()` method.
+
+```js
+E('EXAMPLE_KEY1', 'This is the error value');
+E('EXAMPLE_KEY2', (a, b) => return `${a} ${b}`);
+```
+
+The first argument passed to `E()` is the static identifier. The second
```

Comment:
Yes, but this is a guide so I think this would be read by contributors, especially those who are new? And they might need to change that file and call `E` to register new errors.